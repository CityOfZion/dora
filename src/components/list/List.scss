.data-list-container {
  img {
    max-height: 24px;
  }
}

.data-list > a,
.data-list > #non-link-list-cell-container {
  padding: 0px 24px;
  font-size: 14px;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  min-height: 45px;
  background-color: var(--dark-grey);
  display: flex;
  align-items: center;
  text-decoration: none;
  color: var(--white);
}

.block-index-cell {
  color: var(--pink);
  letter-spacing: 0.96px;
}

.txid-index-cell {
  color: var(--pink);
  max-width: 160px;
  text-overflow: ellipsis;
  overflow: hidden;
  font-size: 14px;
  color: #d355e7;
  letter-spacing: 0;
}

.data-list-column {
  padding: 0px 24px;
  font-weight: 300;
  display: flex;
  align-items: center;
  font-weight: 600;
  font-size: 14px;
  background-color: var(--medium-grey-blue);
  min-height: 25px;
}

.data-list {
  display: grid;
  grid-row-gap: 8px;
}

.loading-table-row {
  opacity: 0.75;
}

.without-pointer-cursor {
  cursor: default !important;
}

.list-header-and-pagination-info-row {
  display: flex;
  align-items: center;
  justify-content: space-between;
  margin-bottom: 12px;
}

// TODO: implement a solution that doesnt cause the list component to rerender
// a:hover {
//   opacity: 0.75;
// }
// .cellhovered img {
//   filter: invert(0%);
// }
// .cellhovered {
// }
.header-cell-container {
  cursor: pointer;
}
.data-list-count-stats {
  justify-content: flex-end;
  display: flex;
  width: 100%;
  margin-bottom: 12px;
}

.data-list-arrow-sort {
  background-color: inherit;
  border: none;
  margin-left: 2px;
  cursor: pointer;
}

@media screen and (max-width: 768px) {
  .data-list > span,
  .data-list-column,
  .txid-index-cell {
    font-size: 11px;
  }
}
