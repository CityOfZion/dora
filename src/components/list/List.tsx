import React, { ReactText } from 'react'
import uniqueId from 'lodash/uniqueId'
import classNames from 'classnames'
import { ReactComponent as ArrowSortSVG } from '../../assets/icons/arrow-sort.svg'
import { SORT_OPTION } from '../../reducers/nodeReducer'
import './List.scss'

interface HeaderCell {
  styleHeader?: React.CSSProperties
  classNameHeader?: string
  nameColumn?: React.Key | null
  isLoading?: boolean
  orderData?: boolean
  sortOpt?: SORT_OPTION
  callbalOrderData?: (field: SORT_OPTION) => void
}
const HeaderCell: React.FC<HeaderCell> = ({
  styleHeader,
  classNameHeader,
  nameColumn,
  isLoading,
  orderData,
  sortOpt,
  callbalOrderData,
}) => {
  return (
    <div
      style={styleHeader}
      className={`${classNameHeader} header-cell-container`}
      key={nameColumn}
      onClick={(e): void => {
        e.preventDefault()
        callbalOrderData && sortOpt && callbalOrderData(sortOpt)
      }}
    >
      {isLoading ? '' : nameColumn}
      {orderData ? (
        <div className="data-list-arrow-sort">
          <ArrowSortSVG />
        </div>
      ) : (
        <></>
      )}
    </div>
  )
}

export type ColumnType = {
  name: string
  accessor: string
  style?: {}
  sortOpt?: SORT_OPTION
}

type ListProps = {
  columns: Array<ColumnType>
  // eslint-disable-next-line
  // @ts-ignore
  data: Array<{ [key: string]: string | number | React.FC<{}>; href?: string }>
  handleRowClick?: (data: {
    [key: string]: string | number | React.FC<{}>
  }) => void
  generateHref?: (data: {
    [key: string]: string | number | React.FC<{}>
  }) => string
  isLoading: boolean
  rowId: string
  withoutPointer?: boolean
  leftBorderColorOnRow?:
    | string
    | ((
        id: string | number | void | React.FC<{}>,
        chain?: string | number | React.FC<{}>,
      ) => string)

  countConfig?: {
    total?: number
    label: string
  }
  orderData?: boolean
  callbalOrderData?: (nameColumn: SORT_OPTION) => void
  paddingCell?: { paddingValue: string; indexesColumns: number[] }
}

export const List: React.FC<ListProps> = ({
  columns,
  data,
  handleRowClick,
  generateHref,
  isLoading,
  rowId,
  withoutPointer = false,
  leftBorderColorOnRow = '',
  countConfig,
  orderData,
  callbalOrderData,
  paddingCell,
}) => {
  const sortedByAccessor = data.map(data => {
    interface Sorted {
      id: string
      href: string
      [key: string]: string | number | React.FC<{}>
    }

    const sorted = {} as Sorted
    columns.forEach(column => {
      sorted[column.accessor] = data[column.accessor]
      sorted.id = String(data[rowId])
      sorted.href = data.href || '#'
      sorted.chain = data.chain
    })
    return sorted
  })

  const gridstyle = {
    gridTemplateColumns: `repeat(${columns.length}, auto)`,
  }

  const conditionalBorderRadius = (
    index: number,
    shouldReturnBorderLeftStyle?: boolean,
    id?: string | number | React.FC<{}>,
    chain?: string | number | React.FC<{}>,
  ): { borderRadius: string } | undefined => {
    if (!index) {
      const border = {
        borderRadius: '3px 0 0 3px',
        borderLeft: '',
      }
      if (shouldReturnBorderLeftStyle && leftBorderColorOnRow) {
        border.borderLeft = `solid 3px ${
          typeof leftBorderColorOnRow === 'function'
            ? leftBorderColorOnRow(id, chain)
            : leftBorderColorOnRow
        }`
      }
      return border
    }
    if (index === columns.length) {
      const border = {
        borderRadius: '0 3px 3px 0',
      }
      return border
    }
    return undefined
  }

  const adjustPaddingStyle = (
    index: number,
  ):
    | {
        padding: string
      }
    | undefined => {
    if (paddingCell) {
      const verify = paddingCell.indexesColumns.find(
        indexColumn => indexColumn === index,
      )
      if (typeof verify !== 'undefined' && verify >= 0) {
        return {
          padding: paddingCell.paddingValue,
        }
      }
    }
  }

  const rowClass = classNames({
    'loading-table-row': isLoading,
    'without-pointer-cursor': withoutPointer,
  })

  const headerRowClass = classNames({
    'loading-table-row': isLoading,
    'data-list-column': true,
  })

  // const [currentHoveredIndex, setCurrentHoveredIndex] = React.useState(-1)

  const renderCellData = (
    isLoading: boolean,
    data: string | number | React.FC<{}>,
  ): ReactText | React.ReactNode => {
    const cellProps = {}
    if (isLoading) return undefined
    if (typeof data === 'function') return data(cellProps)
    return data
  }

  return (
    <div className="data-list-container">
      {countConfig && (
        <div className="data-list-count-stats">
          {' '}
          {countConfig.label} 1 to {data.length}{' '}
          {!!countConfig.total
            ? `of ${countConfig.total.toLocaleString()}`
            : null}
        </div>
      )}
      <div className="data-list" style={gridstyle}>
        {columns.map((column, i) => {
          return orderData ? (
            <HeaderCell
              classNameHeader={headerRowClass}
              styleHeader={{
                ...conditionalBorderRadius(i),
                ...(column.style || {}),
              }}
              key={column.name}
              nameColumn={column.name}
              isLoading={isLoading}
              orderData={orderData}
              sortOpt={column.sortOpt}
              callbalOrderData={callbalOrderData}
            />
          ) : (
            <div
              style={{
                ...conditionalBorderRadius(i),
                ...(column.style || {}),
              }}
              className={headerRowClass}
              key={column.name}
            >
              {isLoading ? '' : column.name}
            </div>
          )
        })}

        {sortedByAccessor.map(
          (
            data: {
              [key: string]: string | number | React.FC<{}>
            },
            index: number,
          ) =>
            Object.keys(data).map((key, i) => {
              const conditionalHref = (): string => {
                if (typeof data.href === 'string' && data.href !== '#') {
                  return data.href
                }
                if (generateHref) {
                  return generateHref(data)
                }
                return '#'
              }

              return !paddingCell
                ? key !== 'id' &&
                    key !== 'href' &&
                    key !== 'chain' &&
                    // TODO: this should probably be using the <Link/> component
                    (typeof data.href === 'string' || generateHref ? (
                      <a
                        href={conditionalHref()}
                        style={conditionalBorderRadius(
                          i,
                          true,
                          data.id,
                          data.chain,
                        )}
                        onClick={(): void =>
                          handleRowClick && handleRowClick(data)
                        }
                        key={uniqueId()}
                        className={rowClass}
                      >
                        {renderCellData(isLoading, data[key])}
                      </a>
                    ) : (
                      <div
                        id="non-link-list-cell-container"
                        style={conditionalBorderRadius(
                          i,
                          true,
                          data.id,
                          data.chain,
                        )}
                        onClick={(): void =>
                          handleRowClick && handleRowClick(data)
                        }
                        key={uniqueId()}
                        className={rowClass}
                      >
                        {renderCellData(isLoading, data[key])}
                      </div>
                    ))
                : key !== 'id' &&
                    key !== 'href' &&
                    key !== 'chain' &&
                    // TODO: this should probably be using the <Link/> component
                    (!data.href.toString().startsWith('#') || generateHref ? (
                      <a
                        href={conditionalHref()}
                        style={{
                          ...conditionalBorderRadius(
                            i,
                            true,
                            data.id,
                            data.chain,
                          ),
                          ...adjustPaddingStyle(i),
                        }}
                        onClick={(): void =>
                          handleRowClick && handleRowClick(data)
                        }
                        key={uniqueId()}
                        className={rowClass}
                      >
                        {renderCellData(isLoading, data[key])}
                      </a>
                    ) : (
                      <div
                        id="non-link-list-cell-container"
                        style={conditionalBorderRadius(
                          i,
                          true,
                          data.id,
                          data.chain,
                        )}
                        onClick={(): void =>
                          handleRowClick && handleRowClick(data)
                        }
                        key={uniqueId()}
                        className={rowClass}
                      >
                        {renderCellData(isLoading, data[key])}
                      </div>
                    ))
            }),
        )}
      </div>
    </div>
  )
}

export default List
