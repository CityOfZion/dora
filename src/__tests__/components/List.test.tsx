import React, { ReactElement } from 'react'
import { shallow } from 'enzyme'

import List from '../../components/list/List'

const columns = [
  {
    name: 'Test Label',
    accessor: 'test',
    style: { backgroundColor: 'purple' },
  },
  {
    name: 'Test Label 2',
    accessor: 'test2',
    style: { backgroundColor: 'purple' },
  },
]

const data = [
  {
    test: 123423445,
    test2: (): ReactElement => <div> hello test 2 </div>,
  },
  {
    test: 123423445,
    test2: (): ReactElement => <div> hello test 2 </div>,
  },
  {
    test: 123423445,
    test2: (): ReactElement => <div> hello test 2 </div>,
  },
  {
    test: 123423445,
    test2: (): ReactElement => <div> hello test 2 </div>,
  },
  {
    test: 123423445,
    test2: (): ReactElement => <div> hello test 2 </div>,
  },
]

test('renders without crashing', () => {
  const tree = shallow(
    <List
      columns={columns}
      data={data}
      rowId={'test'}
      handleRowClick={(row: {
        [key: string]: string | number | React.FC<{}>
      }): void => console.log(row)}
      isLoading={false}
    />,
  )
  expect(tree).toMatchSnapshot()
})
