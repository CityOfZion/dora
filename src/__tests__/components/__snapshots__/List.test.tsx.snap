// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  className="data-list-container"
>
  <div
    className="data-list"
    style={
      Object {
        "gridTemplateColumns": "repeat(2, auto)",
      }
    }
  >
    <div
      className="data-list-column"
      key="Test Label"
      style={
        Object {
          "backgroundColor": "purple",
          "borderLeft": "",
          "borderRadius": "3px 0 0 3px",
        }
      }
    >
      Test Label
    </div>
    <div
      className="data-list-column"
      key="Test Label 2"
      style={
        Object {
          "backgroundColor": "purple",
        }
      }
    >
      Test Label 2
    </div>
    <a
      className=""
      href="#"
      key="1"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      style={
        Object {
          "borderLeft": "",
          "borderRadius": "3px 0 0 3px",
        }
      }
    >
      123423445
    </a>
    <a
      className=""
      href="#"
      key="2"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div>
         hello test 2 
      </div>
    </a>
    <a
      className=""
      href="#"
      key="3"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      style={
        Object {
          "borderLeft": "",
          "borderRadius": "3px 0 0 3px",
        }
      }
    >
      123423445
    </a>
    <a
      className=""
      href="#"
      key="4"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div>
         hello test 2 
      </div>
    </a>
    <a
      className=""
      href="#"
      key="5"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      style={
        Object {
          "borderLeft": "",
          "borderRadius": "3px 0 0 3px",
        }
      }
    >
      123423445
    </a>
    <a
      className=""
      href="#"
      key="6"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div>
         hello test 2 
      </div>
    </a>
    <a
      className=""
      href="#"
      key="7"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      style={
        Object {
          "borderLeft": "",
          "borderRadius": "3px 0 0 3px",
        }
      }
    >
      123423445
    </a>
    <a
      className=""
      href="#"
      key="8"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div>
         hello test 2 
      </div>
    </a>
    <a
      className=""
      href="#"
      key="9"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      style={
        Object {
          "borderLeft": "",
          "borderRadius": "3px 0 0 3px",
        }
      }
    >
      123423445
    </a>
    <a
      className=""
      href="#"
      key="10"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div>
         hello test 2 
      </div>
    </a>
  </div>
</div>
`;
